File: patterns/associated-items/rules-python/associated_items_semantic.py
from openhab import rule, Registry
from openhab.actions import Semantic
from openhab.triggers import ItemStateChangeTrigger


@rule(triggers=[ItemStateChangeTrigger("SomeSensor")])
class SensorHatEinUpdate:
    def execute(self, module, input):
        event = input.get("event")
        if not event:
            return

        sensor = Registry.getItem(event.getItemName())
        equipment = Semantic.getEquipment(sensor)
        if not equipment:
            self.logger.warn("Kein Equipment fuer " + sensor.getName() + " gefunden")
            return

        members = equipment.getAllMembers()

        # Ansatz: Equipment-Name
        by_name = Registry.getItem(equipment.getName() + "_Status")

        # Ansatz: Item-Typ
        by_type = next((i for i in members if i.getType() == "Switch"), None)

        # Ansatz: Item-Tag
        by_tag = next((i for i in members if "Status" in i.getTags()), None)

        # Ansatz: mehrere Kriterien
        by_multi = next(
            (i for i in members if "Status" in i.getTags() and i.getName().endswith("_Status")),
            None,
        )

        by_name.postUpdate(sensor.getState())

File: patterns/associated-items/rules-js/associated-items-semantic.js
rules.JSRule({
  name: "Sensor hat ein Update",
  triggers: [triggers.ItemStateChangeTrigger("SomeSensor")],
  execute: (event) => {
    const equipment = actions.Semantics.getEquipment(items.getItem(event.itemName));
    if (!equipment) {
      console.warn("Kein Equipment fuer SomeSensor gefunden");
      return;
    }

    // equipment.members ist ein rohes Java-Set (equipment kommt aus einer
    // Java-Action, nicht aus dem JS-Wrapper) - erst in echte JS-Items
    // umwandeln, damit JS-Array-Methoden wie .find()/.includes() nutzbar sind.
    const members = Array.from(equipment.members).map((i) => items.getItem(i.name));

    // Ansatz: Equipment-Name (nur als Beispiel, nicht robust: setzt voraus,
    // dass ein Item exakt "<Equipmentname>_Status" existiert)
    // const byName = items.getItem(equipment.name + "_Status");

    // Ansatz: Item-Typ
    const byType = members.find((i) => i.type === "Switch");

    // Ansatz: Item-Tag
    const byTag = members.find((i) => i.tags.includes("Status"));

    // Ansatz: mehrere Kriterien (robust, unabhaengig vom Equipment-Namen)
    const byMulti = members.find(
      (i) => i.tags.includes("Status") && i.name.endsWith("_Status")
    );

    if (!byMulti) {
      console.warn("Kein Status-Item im Equipment " + equipment.name + " gefunden");
      return;
    }
    byMulti.postUpdate(items.getItem(event.itemName).state);
  }
});
